### PR TITLE
[JENKINS-39355 follow up] Add eventing support to API

### PR DESCRIPTION
**Pull request by stephenc**
I give up tracking what Jenkins issues this sweeps up.

Big improvements to the branch API

- Lots of simplifications to how customizations and branding get applied
- Now we support events from SCM API and optimize reindexing using the event information
- Lots of tests of the event support added
- Some other bugs found and fixed

@reviewbybees

### CODE REVIEW

**Comment by stephenc** [comment]
Current code coverage results (with the new tests)

screen shot 2016-11-18 at 17 40 50

**Comment by stephenc** [comment]
screen shot 2016-11-21 at 13 18 19

That is the coverage of IMHO the critical use cases and major functionality in Branch API

File: src/main/java/jenkins/branch/OrganizationFolder.java [hunk 1]
```diff
@@ -337,5 +619,260 @@ public String getDisplayName() {
...
+    }
+
+    @Extension
+    public static class SCMEventListenerImpl extends SCMEventListener {
+
+        public TaskListener globalEventsListener() {
+            File eventsFile =
+                    new File(Jenkins.getActiveInstance().getRootDir(), OrganizationFolder.class.getName() + ".log");
+            boolean rotate = eventsFile.length() > 30 * 1024;
+            OutputStream os = new ReopenableRotatingFileOutputStream(eventsFile, 5);
```
Comment (jglick) [inline src/main/java/jenkins/branch/OrganizationFolder.java]: We need to introduce some sort of reusable utility for this, not least because it will all need to be rewritten for @oleg-nenashev’s `LoggingMethod` changes.

File: src/main/resources/jenkins/branch/DescriptionColumn/column.jelly [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core">
+  <td>
+    <j:set var="p" value="${col.getPropertyOf(job)}"/>
+    <j:choose>
+      <j:when test="${p != null and p.objectUrl != null}">
+        <a href="${p.objectUrl}">${col.description(p,job)}</a>
```
Comment (jglick) [inline src/main/resources/jenkins/branch/DescriptionColumn/column.jelly]: Are `description`s (and `displayName`s) expected to be plain text or may they contain HTML? The assumption here is that they are plain text.

File: src/main/resources/jenkins/branch/LocalizedViewsTabBar/viewTabs.jelly [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout"
+         xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+  <!-- view tab bar -->
+  <l:tabBar>
+    <j:forEach var="v" items="${it.sort(views)}">
+      <l:tab name="${v.displayName}" active="${v==currentView}" href="${rootURL}/${v.url}"/>
+    </j:forEach>
+    <j:if test="${currentView.hasPermission(currentView.CREATE)}">
+      <l:tab name="+" href="${rootURL}/${currentView.owner.url}newView" active="false"
+             title="${%New View}"/>
```
Comment (jglick) [inline src/main/resources/jenkins/branch/LocalizedViewsTabBar/viewTabs.jelly]: Feels like something that should be defined upstream, I suppose in `cloudbees-folder`.

File: src/test/java/integration/BrandingTest.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+/*
+ * The MIT License
```
Comment (jglick) [inline src/test/java/integration/BrandingTest.java]: dupes (elsewhere too, search for them)

File: src/test/java/integration/harness/MockSCMController.java [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        repo.revisions.put(state.getHash(), state);
+        if (branchName != null) {
+            repo.heads.put(branchName, state.getHash());
+        }
+        if (crNum != null) {
+            repo.changes.put(crNum, state.getHash());
+        }
+    }
+
+    public synchronized void rmFile(String repository, String branchOrCR, String message, String path)
```
Comment (jglick) [inline src/test/java/integration/harness/MockSCMController.java]: This kind of stuff seems like it would be useful elsewhere; `workflow-multibranch` and similar currently use real SCM sources (for example with `GitSampleRepoRule`), but having a mock is sometimes nicer. Do you plan to

```xml
<no-test-jar>false</no-test-jar>
```

?

File: src/test/java/jenkins/branch/MessagesTest.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import org.junit.experimental.theories.Theories;
+import org.junit.experimental.theories.Theory;
+import org.junit.runner.RunWith;
+
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+
+/**
+ * The code coverage for the generated Messages class doesn't matter, so this test stops them from dragging the coverage
+ * down and results in a clearer measure of code coverage for the module.
```
Comment (jglick) [inline src/test/java/jenkins/branch/MessagesTest.java]: die die die

File: src/test/java/jenkins/branch/OrganizationFolderTest.java [hunk 7]
```diff
@@ -154,30 +153,4 @@ public String getDisplayName() {
         }
     }
 
-    private RingBufferLogHandler createJULTestHandler() throws SecurityException, IOException {
```
Comment (jglick) [inline src/test/java/jenkins/branch/OrganizationFolderTest.java]: BTW this is superseded by `LoggerRule` anyway.

File: src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            long startTime = System.currentTimeMillis();
+            assertThat(master.isInQueue(), is(true));
+
+            // while it is in the queue, until queue maintenance takes place, it will not be flagged as blocked
+            // since we cannot know when queue maintenance happens from the periodic task
+            // we cannot assert any value of isBlocked() on this side of maintenance
+            Queue.getInstance().maintain();
+            assertThat(master.getQueueItem().isBlocked(), is(true));
+            assertThat(master.getQueueItem().getCauseOfBlockage().getShortDescription().toLowerCase(),
+ containsString("throttle"));
```
Comment (jglick) [inline src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java]: indentation

File: src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+                            )
+                    )
+            );
+            assertThat(master.isInQueue(), is(false));
+            assertThat(master.getQueueItem(), nullValue());
+            QueueTaskFuture<FreeStyleBuild> future = master.scheduleBuild2(0);
+
+            // let the item get added to the queue
+            while (!master.isInQueue()) {
+                Thread.yield();
```
Comment (jglick) [inline src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java]: `waitUntilNoActivity`?

File: src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+            assertThat(master.getQueueItem().getCauseOfBlockage().getShortDescription().toLowerCase(),
+ containsString("throttle"));
+
+            // now we wait for the start... invoking queue maintain every 100ms so that the queue
+            // will pick up more responsively than the default 5s
+            Future<FreeStyleBuild> startCondition = future.getStartCondition();
+            long endTime = startTime + 60*60/rate*1000L*5; // at least 5 times the expected delay
+            while (!startCondition.isDone() && System.currentTimeMillis() < endTime) {
+                Queue.getInstance().maintain();
+                Thread.sleep(100);
```
Comment (jglick) [inline src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java]: ditto

File: src/main/java/jenkins/branch/MultiBranchProjectDescriptor.java [hunk 11]
```diff
@@ -56,6 +60,83 @@
...
+        projectClass = inferProjectClass(clazz);
+    }
+
+    /**
+     * Fully inferring constructor to use when the descriptor is an inner class of the {@link MultiBranchProject}
+     * and type parameter inference works because it just should work.
+     *
+     * @since 2.0
+     */
+    protected MultiBranchProjectDescriptor() {
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProjectDescriptor.java]: Again please make sure you test against: `MatrixMultiBranchProject`, `MavenMultiBranchProject`, `FreeStyleMultiBranchProject`, `IvyMultiBranchProject`, `YamlMultiBranchProject`, `PipelineMultiBranchDefaultsProject`, `FreeStyleMultiBranchProject`, `WorkflowMultiBranchProject`, `LiterateMultibranchProject`

File: src/main/java/jenkins/branch/OrganizationFolder.java [hunk 12]
```diff
@@ -875,4 +1031,125 @@ public void addAttribute(@NonNull String key, Object value)
...
+        /**
+         * Save the settings to a file.
+         */
+        @Override
+        public synchronized void save() throws IOException {
+            if (BulkChange.contains(this)) {
+                return;
+            }
+            getStateFile().write(this);
+            SaveableListener.fireOnChange(this, getStateFile());
```
Comment (jglick) [inline src/main/java/jenkins/branch/OrganizationFolder.java]: again fire outside block

File: src/main/java/jenkins/branch/ParameterDefinitionBranchProperty.java [hunk 13]
```diff
@@ -25,43 +25,82 @@
...
-    /** Not to be confused with {@link BranchPropertyDescriptor#isApplicable(MultiBranchProjectDescriptor)}, this checks applicability for the child job type. */
+    /**
+     * Tests if the parameter definitions are applicable to the branch specific job type.
+     * <p>
+     * Not to be confused with {@link BranchPropertyDescriptor#isApplicable(MultiBranchProjectDescriptor)},
+     * this checks applicability for the child job type.
+     * <p>
+     * As all {@link Job} types support {@link JobProperty} and {@link ParametersDefinitionProperty} will work
+     * with anything that implements {@link ParameterizedJobMixIn.ParameterizedJob}, the default implementation
+     * which checks for both of these interfaces, should suffice for all. If you need to appliy additiona specification
```
Comment (jglick) [inline src/main/java/jenkins/branch/ParameterDefinitionBranchProperty.java]: typos

File: src/main/java/jenkins/branch/ParameterDefinitionBranchProperty.java [hunk 14]
```diff
@@ -25,43 +25,82 @@
...
+     * @param <P>   the type of the branch specific child job.
+     * @param <B>   the type of the branch specific child job's builds.
+     * @param clazz the type of the branch specific child job.
+     * @return {@code true} if the specified type of child job can be parameterized.
+     */
+    @OverrideMustInvoke
     protected <P extends Job<P, B>, B extends Run<P, B>> boolean isApplicable(Class<P> clazz) {
-        return true;
+        return Job.class.isAssignableFrom(clazz)
+                && ParameterizedJobMixIn.ParameterizedJob.class.isAssignableFrom(clazz);
```
Comment (jglick) [inline src/main/java/jenkins/branch/ParameterDefinitionBranchProperty.java]: Note that this is only implemented in plugins which request it anyway—deliberately not offered for Pipeline.

File: src/main/java/jenkins/branch/RateLimitBranchProperty.java [hunk 15]
```diff
@@ -351,25 +380,25 @@ public FormValidation doCheckCount(@QueryParameter int value, @QueryParameter St
...
                 if (interval < TimeUnit2.DAYS.toMillis(14)) {
                     return FormValidation.ok(
-                                    Messages.RateLimitBranchProperty_ApproxDaysBetweenBuilds(
-                                            TimeUnit2.MILLISECONDS.toDays(interval)));
+                            Messages.RateLimitBranchProperty_ApproxDaysBetweenBuilds(
+                                    TimeUnit2.MILLISECONDS.toDays(interval)));
                 }
                 return FormValidation.ok(Messages.RateLimitBranchProperty_ApproxWeeksBetweenBuilds(
-                                TimeUnit2.MILLISECONDS.toDays(interval) / 7));
+                        TimeUnit2.MILLISECONDS.toDays(interval) / 7));
```
Comment (jglick) [inline src/main/java/jenkins/branch/RateLimitBranchProperty.java]: Please stop reformatting!

File: src/main/java/jenkins/branch/RateLimitBranchProperty.java [hunk 16]
```diff
@@ -386,21 +415,60 @@ public FormValidation doCheckCount(@QueryParameter int value, @QueryParameter St
...
+                                            betweenBuilds
+                                    }
+                            );
                             return CauseOfBlockage.fromMessage(Messages._RateLimitBranchProperty_BuildBlocked(
                                     new Date(lastBuild.getTimeInMillis() + betweenBuilds))
                             );
                         }
                     }
+                    // ensure items leave the queue in the order they were scheduled
+                    List<Queue.Item> items = Jenkins.getActiveInstance().getQueue().getItems(item.task);
```
Comment (jglick) [inline src/main/java/jenkins/branch/RateLimitBranchProperty.java]: How is this change related to anything else in this PR? Should have been filed separately with an explanation.

File: src/test/java/jenkins/branch/CustomOrganizationFolderDescriptorTest.java [hunk 17]
```diff
@@ -31,21 +31,16 @@
...
 import java.util.Collections;
 import java.util.List;
 import java.util.logging.Level;
 import jenkins.scm.api.SCMNavigator;
 import jenkins.scm.api.SCMNavigatorDescriptor;
 import jenkins.scm.api.SCMSourceObserver;
 import jenkins.scm.impl.SingleSCMNavigator;
-
-import static org.hamcrest.Matchers.containsInAnyOrder;
-import static org.junit.Assert.*;
```
Comment (jglick) [inline src/test/java/jenkins/branch/CustomOrganizationFolderDescriptorTest.java]: Gratuitous; use of static star imports is customary for such methods.

File: src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java [hunk 18]
```diff
@@ -160,7 +174,95 @@ public void rateLimitsBlockBuilds(int rate) throws Exception {
...
+                            lessThanOrEqualTo(60 * 60 / rate * 1000L * 500L)
+                    )
+            );
+            future.get();
+        }
+
+    }
+
+    @Test
+    public void rateLimitsConcurrentBuilds() throws Exception {
```
Comment (jglick) [inline src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java]: Seems unrelated to the PR as a whole.

File: src/main/java/jenkins/branch/RateLimitBranchProperty.java (hunk 16 above)
Comment (stephenc) [inline src/main/java/jenkins/branch/RateLimitBranchProperty.java]: It fell out of tests being added for event support. I had a question about whether the rate limit would work with events, then when I wrote up a test I found a bug in the rate limit and then I discovered that the bug was generic and I didn't bother resurrecting the event version of the test as it the non-event version is simpler (and not flakey)

File: src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java (hunk 18 above)
Comment (stephenc) [inline src/test/java/jenkins/branch/RateLimitBranchPropertyTest.java]: only because I removed the event based test that found the bug about concurrent entries in the queue

File: src/main/java/jenkins/branch/MetadataActionFolderIcon.java [hunk 19]
```diff
@@ -89,7 +89,11 @@ public String getIconClassName() {
                         return null;
                     }
                 } catch (IllegalArgumentException ignore) {
-                    // should never happen, but if it does we will just fall back to icon class name
+                    // This would be thrown by Util.isOverridden if the method does not exist, which
+                    // should never happen, as AvatarMetadataAction has the method (which we call)
```
Comment (jglick) [inline src/main/java/jenkins/branch/MetadataActionFolderIcon.java]: So

```java
assert false : ignore;
```

just in case you are wrong.

File: src/main/java/jenkins/branch/MultiBranchProjectViewHolder.java [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+     * A custom category specific view.
+     */
+    @Restricted(NoExternalUse.class)
+    public static class ViewImpl extends ListView {
+
+        /**
+         * The category that this view selects.
+         */
+        @NonNull
+        private final SCMHeadCategory category;
```
Comment (KostyaSha) [inline src/main/java/jenkins/branch/MultiBranchProjectViewHolder.java]: field is hidden, impossible to get category for the given view

Comment (stephenc) [reply src/main/java/jenkins/branch/MultiBranchProjectViewHolder.java]: Pr for getter welcome. Quit sniping, file PRs to improve if you think they are required, i’ll Discuss potential changes, but sniping on random commits is not winning you friends here
